dali/pipeline/graph/op_graph2.cc

```diff
@@ -0,0 +1,313 @@
+  } else if (device == "cpu") {
+    return OpType::CPU;
+  } else if (device == "mixed") {
+    return OpType::MIXED;
+  }
+  DALI_FAIL("Unsupported device type: " + device + ".");
+}
+
+StorageDevice ParseStorageDevice(const std::string &io_device) {
+  if (io_device == "cpu") {
+    return StorageDevice::CPU;
+  }
+  return StorageDevice::GPU;
+}
+
+}  // namespace
+
+//////////////////////////////////////////////////////////////////////////////
+// OpGraph
+
+OpNode &OpGraph::AddOp(std::string instance_name, OpSpec spec) {
+  OpNodeList tmp;
+  auto &op_node = tmp.emplace_back(std::move(instance_name), std::move(spec));
+  if (!name2op_.emplace(op_node.instance_name, &op_node).second) {
+    throw std::invalid_argument(
+        make_string("Duplicate operator instance name: \"", op_node.instance_name, "\""));
+  }
+  op_node.iter = tmp.begin();
+  op_nodes_.splice(op_nodes_.end(), tmp);
+  return op_node;
+}
+
+DataNode &OpGraph::AddData(std::string name, StorageDevice device) {
+  DataNodeList tmp;
+  auto &data_node = tmp.emplace_back(std::move(name), device);
+  if (!name2data_.emplace(data_node.name, &data_node).second) {
+    throw std::invalid_argument(
+        make_string("Duplicate data node name: \"", data_node.name, "\""));
+  }
+  data_node.iter = tmp.begin();
+  data_nodes_.splice(data_nodes_.end(), tmp);
+  return data_node;
+}
+
+void OpGraph::RemoveDataNodeReferences(OpNode &op) {
+  for (auto *data : op.inputs) {
+    auto it = std::remove_if(data->consumers.begin(), data->consumers.end(), [&](DataEdge e) {
+      return e.op == &op;
+    });
+    // the removed op must have been a consumer
+    assert(it != data->consumers.end());
+    data->consumers.erase(it, data->consumers.end());
+  }
+  for (auto *data : op.outputs) {
+    // the removed op must have been the producer
+    assert(data->producer.op == &op);
+    data->producer = {};
+  }
+}
+
+
+bool OpGraph::EraseOp(std::string_view name) {
+  auto it = name2op_.find(name);
+  if (it == name2op_.end())
+    return false;
+  auto &op = *it->second;
+
+  RemoveDataNodeReferences(op);
+
+  op_nodes_.erase(it->second->iter);
+  name2op_.erase(it);
+  return true;
+}
+
+
+bool OpGraph::EraseData(std::string_view name) {
+  auto it = name2data_.find(name);
+  if (it == name2data_.end())
+    return false;
+
+  auto &node = *it->second;
+  if (node.producer.op) {
+    assert(node.producer.op->outputs.size() >= static_cast<size_t>(node.producer.idx));
+    assert(node.producer.op->outputs[node.producer.idx] == &node);
+    node.producer.op->outputs[node.producer.idx] = nullptr;
+  }
+
+  for (auto &edge : node.consumers) {
+    if (edge.op == nullptr)
+      continue;
+    assert(edge.op->inputs.size() >= static_cast<size_t>(edge.idx));
+    assert(edge.op->inputs[edge.idx] == &node);
+    edge.op->inputs[edge.idx] = nullptr;
+  }
+
+  data_nodes_.erase(it->second->iter);
+  name2data_.erase(it);
+  return true;
+}
+
+int OpGraph::AddOutput(std::string_view name) {
```

Comment:
It will. It just stores a view of the output's name in a vector and marks the relevant datanode as output. Adding an output twice will make it appear twice in the vector, but it won't make the output flag "truer".